# LAB 02.03 - Pandas

In [1]:
!wget --no-cache -O init.py -q https://raw.githubusercontent.com/rramosp/ai4eng.v1/main/content/init.py
import init; init.init(force_download=False); init.get_weblink()


replicating local resources


In [2]:
from local.lib.rlxmoocapi import submit, session
session.LoginSequence(endpoint=init.endpoint, course_id=init.course_id, lab_id="L02.03", varname="student");

logging in as andres.quintero8@udea.edu.co ... please wait
using course session ai4eng.v1.udea.r2::2022.1
success!! you are logged in


In [3]:
import numpy as np
import pandas as pd

## Task 1: Extract data

we have a dataframe of items with a price and, **sometimes**, with an extra column (`margin`).

You will have to complete a function that will **filter** the dataframe selecting the rows:

- whose price is > 100
- **OR** whose margin is >10, if the column `margin` is present

your function must return **A LIST** with the item ids of the selected rows

For instance, with the following dataframe

             price  category
    itemid                  
    56556   108.15       2.0
    73065    83.54       2.0
    36619   114.42       0.0
    73414    82.94       0.0
    13410   115.13       2.0
    66153    91.50       1.0
    77380    85.82       0.0
    73249    95.95       0.0
    11654   100.22       2.0
    11972    77.16       1.0

your must return the following list:

     [56556, 36619, 13410, 11654]
     
But if you get the following dataframe

             price  category  margin
    itemid                          
    39059    98.11       0.0   11.04
    19526    98.11       1.0   11.25
    78176    94.34       1.0   10.51
    50948   102.37       1.0   10.77
    12111    98.07       1.0    8.50
    56191    98.53       1.0   11.65
    38887    91.49       2.0   11.24
    77915   117.30       0.0    8.64
    55010    96.13       0.0    8.95
    45925    98.59       1.0   10.45
    
you must return the following list

    [39059, 19526, 78176, 50948, 56191, 38887, 77915, 45925]

In [4]:
def create_df(missing=False, n=10):
    itemid   = np.random.randint(100000, size=n)+1000
    category = np.random.randint(3, size=n)
    price    = np.round(np.random.normal(loc=100, scale=10, size=n),2)
    margin   = np.round(np.random.normal(loc=10, scale=1, size=n),2)
    
    if missing:
        nmissing = np.random.randint(len(price)//2)+2                                     
        price[np.random.permutation(len(price))[:nmissing]] = np.nan
    
    d = pd.DataFrame(np.r_[[price, category, margin]].T, index=itemid, columns=["price", "category", "margin"])
    d.index.name="itemid"
    if np.random.random()>.5:
        d = d[d.columns[:2]]
        
    return d

In [5]:
d = create_df()
d

price  category
itemid                  
60067   106.37       2.0
14463    90.26       2.0
55574   111.28       0.0
26314   106.47       1.0
85411    96.52       1.0
63934   112.42       1.0
71439   101.38       2.0
13916    79.86       2.0
47379    72.29       0.0
6595    115.36       0.0

In [6]:
def select_items(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()
    
    names = list(df.columns.values)
    longitud=len(names)
    
    if longitud == 2:
      result= list(np.array((df[(df['price'] > 100)]).index))
      return result
    else:
      result= list(np.array((df[((df['price'] > 100) | (df['margin'] > 10))]).index))
    return result

**manually check your answer**

In [7]:
d = create_df()
d

price  category  margin
itemid                          
45225   118.71       2.0   10.45
69425    99.71       1.0   10.19
61284   100.41       2.0    8.65
18022   105.31       2.0   11.76
51508   112.41       1.0    8.77
91581   103.82       2.0    9.88
59082   105.26       2.0    9.91
71691    95.51       1.0   11.05
53027    99.17       0.0    9.64
47282    79.57       1.0   10.01

In [8]:
select_items(d)

[45225, 69425, 61284, 18022, 51508, 91581, 59082, 71691, 47282]

**submit your code**

In [9]:
student.submit_task(globals(), task_id="task_01");

## Task 2: Group statistics

Complete the following function so that it returns a dataframe with the average, max and min **prices** per category. 

For instance, for the following dataframe

             price  category  margin
    itemid                          
    17946    93.85       1.0   10.64
    61190    91.72       1.0    9.76
    39639   100.16       1.0   10.67
    17791   110.44       2.0    9.65
    7333    101.05       1.0    9.69
    77362   122.33       0.0   11.14
    92646   108.13       2.0   10.58
    27797    85.52       2.0   10.88
    31746    97.56       0.0    9.75
    12355   101.04       2.0    9.51
    
you should return the following dataframe

                 media  maximo  minimo
    categoria                         
    0         109.9450  122.33  97.56
    1          96.6950  101.05  91.72
    2         101.2825  110.44  85.52
    
observe that your result
- must not be a **multilevel** columnset.
- the column names and the index name must be **exactly** as in the example.
- the **index** must be of type **int**.

In [19]:
def get_stats(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()
        
    ... # YOUR CODE HERE
    dfa = pd.DataFrame()

    # creamos las columnas
    dfa['media'] = None
    dfa['maximo'] = None
    dfa['minimo'] = None
    dfa.index.name="categoria"

    Pmean_TZ = np.mean(df[(df['category'] == 0)])
    PMEZ=Pmean_TZ[0]
    Pmean_TO = np.mean(df[(df['category'] == 1)])
    PMEO=Pmean_TO[0]
    Pmean_TT = np.mean(df[(df['category'] == 2)])
    PMET=Pmean_TT[0]
    Pmax_TZ = np.max(df[(df['category'] == 0)])
    PMAZ=Pmax_TZ[0]
    Pmax_TO = np.max(df[(df['category'] == 1)])
    PMAO=Pmax_TO[0]
    Pmax_TT = np.max(df[(df['category'] == 2)])
    PMAT=Pmax_TT[0]
    Pmin_TZ = np.min(df[(df['category'] == 0)])
    PMIZ=Pmin_TZ[0]
    Pmin_TO = np.min(df[(df['category'] == 1)])
    PMIO=Pmin_TO[0]
    Pmin_TT = np.min(df[(df['category'] == 2)])
    PMIT=Pmin_TT[0]
    F1 = 0
    F2 = 1
    F3 = 2
    # añadimos filas por su nombre de fila
    dfa.loc[F1] = [PMEZ, PMAZ, PMIZ]
    dfa.loc[F2] = [PMEO, PMAO, PMIO]
    dfa.loc[F3] = [PMET, PMAT, PMIT]
    result = dfa
    return  result

**manually check your answer**

In [17]:
d = create_df()
d

price  category  margin
itemid                          
52372   105.21       0.0    8.73
23037   114.99       2.0    8.15
72459   102.89       1.0   10.35
43063   106.83       0.0   12.41
80038   108.56       2.0    8.86
38658    99.33       1.0    8.88
72252   105.33       1.0    9.75
24977    80.76       1.0   10.72
6959    103.36       2.0   10.66
7323    103.47       0.0    9.22

In [18]:
get_stats(d)

media  maximo  minimo
categoria                          
0          105.1700  106.83  103.47
1           97.0775  105.33   80.76
2          108.9700  114.99  103.36

**submit your code**

In [20]:
student.submit_task(globals(), task_id="task_02");

## Task 3: Fill in missing data

Fill in the missing data in the **price** column with the following procedure:

- compute the mean and std of the avaialable prices
- sample from a normal distribution with the computed mean and std (see [`np.random.normal`])(https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html), as many samples as missing data
- substitute the missing values with the samples

For instance, for this input dataframe:

             price  category  margin
    itemid                          
    18922      NaN       1.0   10.32
    69500   121.25       1.0   10.22
    76442    90.25       1.0   12.60
    33863   106.51       0.0   10.26
    15904    95.87       1.0   11.51
    41946   103.47       2.0    9.85
    85451    93.08       2.0    9.56
    70028   116.68       1.0    9.11
    26860      NaN       2.0    9.71
    12807    91.48       0.0    9.77
    
your solution might be similar to this (not exactly the same as you will be sampling data):


                 price  category  margin
    itemid                              
    18922    97.441188       1.0   10.32
    69500   121.250000       1.0   10.22
    76442    90.250000       1.0   12.60
    33863   106.510000       0.0   10.26
    15904    95.870000       1.0   11.51
    41946   103.470000       2.0    9.85
    85451    93.080000       2.0    9.56
    70028   116.680000       1.0    9.11
    26860   103.294843       2.0    9.71
    12807    91.480000       0.0    9.77
    
    
**WARN**: your function must not modify the original dataframe, make a copy of the input dataframe, fill the values in the copy and return it.

**HINT**: use the [`isna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isna.html) method of a dataframe or a series.

In [ ]:
d = create_df(missing=True)
d

In [21]:
def fillna(df):
  # make sure to make a copy in case you modify the original df
  df = df.copy()
  norms = np.random.normal(loc= np.mean(df.price), scale = np.std(df.price), size=np.sum(df.isna().values))
  df.price[df.price.isna()==True] = norms
  return df

**check manually your code**

In [22]:
d = create_df(missing=True)
d

price  category
itemid                 
30132   93.24       2.0
72168     NaN       1.0
64763   87.27       0.0
21636   99.73       2.0
58189     NaN       2.0
91676     NaN       0.0
42777     NaN       1.0
99874     NaN       0.0
68093     NaN       0.0
18687   82.89       1.0

In [23]:
fillna(d)

price  category
itemid                     
30132   93.240000       2.0
72168   82.189986       1.0
64763   87.270000       0.0
21636   99.730000       2.0
58189   81.189602       2.0
91676   90.293190       0.0
42777   94.685172       1.0
99874   93.003488       0.0
68093   79.088553       0.0
18687   82.890000       1.0

**submit your code**

In [24]:
student.submit_task(globals(), task_id="task_03");